### ISSUE TITLE
[Suggestion]: .NET Core

### ISSUE THREAD

**Issue by Wagnerp** [Wagnerp/Krypton-NET-5.470#237]
I was wondering how feasible it would be to migrate to .NET Core, as this seems to be the future?

**Comment by Wagnerp** [comment]
**This maybe helpful!** https://docs.microsoft.com/en-us/dotnet/core/porting/

**Comment by Smurf-IV** [comment]
I'm not sure how all the PInvoke calls are going to be covered ?

**Comment by Smurf-IV** [comment]
![image](https://user-images.githubusercontent.com/2418812/70849426-11184980-1e76-11ea-822a-7aeb4a0a229e.png)

**Comment by Smurf-IV** [comment]
@Wagnerp Do you want this to happen as a separate repo, or as an extension of 480 ?
Reason being, is that I might have an active project I can use as a test for this going forward, and obviously will need to branch (And fix) from somewhere :-)

**Comment by Smurf-IV** [comment]
Windows Forms Controls Removal
The following Windows Forms controls have been removed from .NET Core 3.1:
    DataGrid
    ToolBar
    ContextMenu
    Menu
    MainMenu
    MenuItem

On A side note:
> To improve Windows desktop stacks and enable .NET desktop developers to benefit from all the updates of the future, we brought Windows Forms and WPF to .NET Core. They will still remain Windows-only technologies because there are tightly coupled dependencies to Windows APIs. But .NET Core, besides being cross-platform, has many other features that can enhance desktop applications.

So should not need to worry about the `PInvoke`s not being there for Linux as the `WinForm` side is for windows only any way !

**Comment by Smurf-IV** [comment]
Seems like there is no "Rush" for the Winforms stuff tho'
>For Windows Forms the runtime part is fully ported to .NET Core and the team is working on the Windows Forms Designer. We are planning to get it ready by the fourth quarter of 2020 and for now you can check out the Preview version of the designer in Visual Studio 16.4 Preview 3 or later.

**Comment by Smurf-IV** [comment]
https://github.com/dotnet/try-convert

**Comment by Smurf-IV** [comment]
https://devblogs.microsoft.com/dotnet/updates-to-net-core-windows-forms-designer-in-visual-studio-16-5-preview-1/?utm_source=vs_developer_news&utm_medium=referral
> We are happy to announce the new preview version of the **.NET Core Windows Forms designer,** which is available with the **Visual Studio 16.5 Preview 1.**
The big news is that the designer is **_now_** part of Visual Studio! This means that installing the .NET Core Windows Forms designer from a separate VSIX is no longer needed!
To use the designer:
You must be using Visual Studio 16.5 Preview 1 or a later version.
You need to enable the designer in Visual Studio. Go to Tools > Options > Environment > Preview Features and select the Use the preview Windows Forms designer for .NET Core apps option.

**Comment by Smurf-IV** [comment]
Are damn it:
> **Features currently under development**
Support for the following features is currently on our backlog and being actively addressed:
Localization of your own Windows Forms applications.
Data-related scenarios including data binding and data-related controls.
Document Outline Window.
The remaining container controls.
MenuStrip and ToolStrip are expected in the next preview.
**_Third-party controls and UserControls._**
**Inherited Forms and UserControls.**
Tab Order.
Tools > Options page for the designer.

So it seems even tho' most of the stuff is in place, the designer part integration of Krypton will still not work !

**Comment by Smurf-IV** [comment]
Ha, I wonder of this hack works for the "Third Party Controls"
>All you need to do is change:
`<TargetFramework>netcoreapp3.0</TargetFramework>`
to
`<TargetFrameworks>net472;netcoreapp3.0</TargetFrameworks>`
And then as always use .net designer ….

**Comment by Wagnerp** [comment]
@Smurf-IV 

Yes, I found that out when experimenting, seems that it's designed for mac/Linux development in mind

**Comment by Wagnerp** [comment]
Problem is that MS is encouraging devs to migrate & more so in .NET 5

**Comment by Smurf-IV** [comment]
.Net Core 3.1 its LTS so it might be worth the effort, so that things are "ready" for .Net 5 when it comes along.
Then it "should" just be another option in the 
`<TargetFrameworks>net472;netcoreapp3.0</TargetFrameworks>` line.

**Comment by Wagnerp** [comment]
Could this be done in this repo, or the [Krypton-Toolkit-Suite-NET-Core](https://github.com/Wagnerp/Krypton-Toolkit-Suite-NET-Core) repo?

**Comment by Smurf-IV** [comment]
Ok, I'll play with the Net-Core repo :-)

**Comment by Smurf-IV** [comment]
@Wagnerp Can you please transfer this issue to "Net-Core", and then create another here and link to the transferred one ? Cheers

**Comment by Wagnerp** [comment]
Hi @Smurf-IV 

Yes of course I can, I also think that I have an idea for future NuGet packages. To put **all** versions into one.

**Comment by Wagnerp** [comment]
I don't know if VS can build for multiple framework versions at once though, as that'll be ideal.

**Comment by Smurf-IV** [comment]
> 
> 
> I don't know if VS can build for multiple framework versions at once though, as that'll be ideal.

That's just it, once converted to the new format it can, and so "All" current versions if the frameworks can be built from the same source, and then they placed into separate directories. Then the Nuget package is modified to download them all (Again different directories), but only attach the ones to the owner project for the compatible framework. Look at how NLog does it.

**Comment by Wagnerp** [comment]
That would be great, as it wouldn't take forever updating the new packages. The question is how to version the overall package, whether to stick to `5.470.xxxx` or `5.500.xxx` to encompass all framework versions?

**Comment by Smurf-IV** [comment]
How about a curve ball until "It's ready" and call it 5.490.xxx :-)
Then when .Net 5 comes out it would follow that name to 5.500
This way we do not break the existing paradim of the numbering, and introduces something that will work for people wanting to upgrade.

**Comment by Wagnerp** [comment]
I've been playing around with NuGet Explorer and it can be done, I do intend to publish a `5.490.2098` version to test with.

**Comment by Wagnerp** [comment]
This maybe useful! https://www.c-sharpcorner.com/article/create-and-publish-a-nuget-package-using-visual-studio-net-standard/

**Comment by Smurf-IV** [comment]
have you create a new Repo, or using an existing one, and or renamed ?

**Comment by Wagnerp** [comment]
I've created a new test repo that contains just the core components from all versions in one solution. I'll invite you, as I've marked it as private.

**Comment by Wagnerp** [comment]
Hi @Smurf-IV 

Bit random but can we use this now in the extended toolkit now it's no longer GPL? https://github.com/grokys/gong-shell

**Comment by Smurf-IV** [comment]
![image](https://user-images.githubusercontent.com/2418812/71408974-e24d6080-2637-11ea-82df-f41626f03008.png)

**Comment by Smurf-IV** [comment]
Each example will need to be run to update small things, because, due to timing, some events are called in a different order and can cause things to throw exceptions.
i.e this
```c#
        private void button_Enter(object sender, EventArgs e)
        {
            // Setup the property grid to edit this button
            propertyGrid.SelectedObject = new KryptonButtonProxy(sender as KryptonButton);
        }
```
become this:
```c#
        private void button_Enter(object sender, EventArgs e)
        {
            // Setup the property grid to edit this button
            if (sender is KryptonButton kb )
                propertyGrid.SelectedObject = new KryptonButtonProxy(kb);
        }
```

otherwise, a null is passed through to the proxy class during initialization.

**Comment by Smurf-IV** [comment]
New Output Layout from a single codebase:
![image](https://user-images.githubusercontent.com/2418812/71512011-4ec09d80-288c-11ea-96e8-551cc9a21fdf.png)
